feat: support Remote Config sampling rules

dmehala opened the pull request.

pablomartinezbernardo commented on src/datadog/config_manager.cpp:
```diff
@@ -35,32 +98,46 @@ std::vector<ConfigMetadata> ConfigManager::update(const ConfigUpdate& conf) {
+    if (auto error = maybe_rules.if_error()) {
+      trace_sampling_rules_metadata.error = std::move(*error);
+    } else {
+      rules.merge(*maybe_rules);
```
I'm looking at [the docs for this](https://en.cppreference.com/w/cpp/container/unordered_map/merge) and if I'm understanding correctly, items already in `rules` will not be overriden by those in `maybe_rules`. Is that the intended behavior?

pablomartinezbernardo reviewed (approved):
One small comment, otherwise LGTM

dmehala commented:
@pablomartinezbernardo 

Since your review I added 3 commits:
- 9545d64206d236f121ae311856d7398a146e1c82
- be97f1986e25e4fde626dde64fa5d9d4c29b139f
- 23bb27cc653b4cb64da4b31b497cc7b6c1c4e941

I fixed the rule override issue you mentioned but also remote configuration system tests revealed part of the implementation I overlooked.

pablomartinezbernardo commented on src/datadog/span_sampler_config.cpp:
```diff
@@ -16,7 +16,12 @@ namespace {
   nlohmann::json res;
   for (const auto &r : rules) {
-    res.emplace_back(r.to_json());
+    auto j = r.to_json();
+    j["sample_rate"] = r.sample_rate;
+    if (r.max_per_second) {
+      j["max_per_second"] = *r.max_per_second;
+    }
+    res.emplace_back(std::move(j));
```
Curious to know, as they are being set in `to_json`, why is it necessary to set them again?

dmehala replied:
`to_json` is not overloaded by `SpanSamplerConfig::Rule`, meaning `SpanMatcher::to_json` is called instead.

pablomartinezbernardo commented on src/datadog/trace_sampler_config.cpp:
```diff
@@ -196,6 +197,9 @@ Expected<FinalizedTraceSamplerConfig> finalize_config(
         ConfigMetadata::Origin::CODE);
+  } else {
+    result.metadata[ConfigName::TRACE_SAMPLING_RATE] = ConfigMetadata(
+        ConfigName::TRACE_SAMPLING_RATE, "1.0", ConfigMetadata::Origin::CODE);
```
Wanted `ConfigMetadata::Origin::DEFAULT` here? Not sure if it has any side effect, and it's true that it is `CODE` as well though

dmehala replied:
Good catch. I did it right but I must have lost that part during a rebase :/

dmehala replied:
```suggestion
        ConfigName::TRACE_SAMPLING_RATE, "1.0", ConfigMetadata::Origin::DEFAULT);
```
